pkg/lib/controller-runtime/source/dynamic.go

```diff
@@ -0,0 +1,196 @@
+package source
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	ctrlcache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("source")
+
+// Dynamic sources events for all kinds exposed in discovery.
+type Dynamic struct {
+	// Injected by the controller manager
+	client discovery.ServerResourcesInterface
+	cache  ctrlcache.Cache
+	stop   <-chan struct{}
+
+	mu      sync.RWMutex
+	sources map[schema.GroupVersionKind]source.Source
+}
+
+var (
+	_ source.Source    = &Dynamic{}
+	_ inject.Cache     = &Dynamic{}
+	_ inject.Stoppable = &Dynamic{}
+)
+
+// Start is internal and should be called only by the Controller to register an EventHandler with the Informer to enqueue reconcile.Requests.
+func (d *Dynamic) Start(handle handler.EventHandler, queue workqueue.RateLimitingInterface, predicates ...predicate.Predicate) error {
+	if d.client == nil {
+		return fmt.Errorf("must inject rest config before starting dynamic source")
+	}
+	if d.cache == nil {
+		return fmt.Errorf("must inject cache before starting dynamic source")
+	}
+	if d.stop == nil {
+		return fmt.Errorf("must inject stop channel before starting dynamic source")
+	}
+
+	// TODO(njhale): Stop sources when GVKs are removed
+	var (
+		synced sync.WaitGroup
+		once   sync.Once
+	)
+	synced.Add(1)
+
+	go wait.Until(func() {
+		defer once.Do(func() {
```

Comment:
The intention here is to only call `synced.Done()` once after the function exits - but since the function is called within the `Until()` scope every 15 seconds are you sure this doesn't just get triggered the first time the function is executed within the `Until` loop? 

This is probably standard code, just want to confirm.

Reply:
> The intention here is to only call synced.Done() once after the function exits - but since the function is called within the Until() scope every 15 seconds are you sure this doesn't just get triggered the first time the function is executed within the Until loop?

I'm not understanding the delineation you're making here. Could you rephrase?